CDP4WebServices.API/Modules/10-25/CherryPickApi.cs

```diff
@@ -0,0 +1,312 @@
+    using CDP4WebServices.API.Services;
+    using CDP4WebServices.API.Services.Authentication;
+    using CDP4WebServices.API.Services.Authorization;
+    using CDP4WebServices.API.Services.Protocol;
+
+    using Nancy;
+    using Nancy.Responses;
+    using Nancy.Security;
+
+    using NLog;
+
+    using Npgsql;
+
+    using Thing = CDP4Common.DTO.Thing;
+
+    /// <summary>
+    /// This is an API endpoint class to support cherry picking of <see cref="CDP4Common.DTO.Thing" /> inside an
+    /// <see cref="Iteration" />
+    /// based on <see cref="Category" /> filter
+    /// </summary>
+    public class CherryPickApi : ApiBase
+    {
+        /// <summary>
+        /// The url to reach this api
+        /// </summary>
+        private const string RouteUrl = "cherrypick/EngineeringModel/{engineeringModelId}/Iteration/{iterationId}";
+
+        /// <summary>
+        /// A <see cref="NLog.Logger" /> instance
+        /// </summary>
+        private static readonly Logger Logger = LogManager.GetCurrentClassLogger();
+
+        /// <summary>
+        /// The supported get query parameters.
+        /// </summary>
+        private static readonly string[] SupportedGetQueryParameters =
+        {
+            QueryParameters.ClassKindQuery,
+            QueryParameters.CategoryQuery
+        };
+
+        /// <summary>
+        /// Initializes a new instance of <see cref="CherryPickApi" />
+        /// </summary>
+        public CherryPickApi()
+        {
+            this.RequiresAuthentication();
+
+            this.Get[RouteUrl] =
+                route => this.GetResponse(route);
+
+            this.Post[RouteUrl] =
+                route => this.PostResponse(route);
+        }
+
+        /// <summary>
+        /// Gets or sets the person resolver service.
+        /// </summary>
+        public IPersonResolver PersonResolver { get; set; }
+
+        /// <summary>
+        /// Gets or sets the obfuscation service.
+        /// </summary>
+        public IObfuscationService ObfuscationService { get; set; }
+
+        /// <summary>
+        /// Abstract method contract for get response data processing.
+        /// </summary>
+        /// <param name="routeParams">
+        /// The route parameters from the request.
+        /// </param>
+        /// <returns>
+        /// The <see cref="Response" />.
+        /// </returns>
+        protected override Response GetResponseData(dynamic routeParams)
+        {
+            NpgsqlConnection connection = null;
+            NpgsqlTransaction transaction = null;
+
+            this.TransactionManager.SetCachedDtoReadEnabled(true);
+
+            var sw = new Stopwatch();
+            sw.Start();
+            var requestToken = this.GenerateRandomToken();
+
+            try
+            {
+                Logger.Info(this.ConstructLog($"{requestToken} started"));
+                HttpRequestHelper.ValidateSupportedQueryParameter(this.Request, this.RequestUtils, SupportedGetQueryParameters);
+
```

Comment:
i think that we should verify that the caller has permissions to request the classkind before we process the request. If the classkind does not have at leas READ, we should exit here and return a 404 not allowed. I want to make sure we do not create a backdoor here.